ephemeral-disk-setup/azure-ephemeral-disk-setup

```diff
@@ -0,0 +1,435 @@
+    fi
+
+    log "SCSI resource disk $disk is NTFS formatted and safe to reformat"
+}
+
+# Read configuration file and export key=value pairs.
+load_config() {
+    if [ ! -r "${CONFIGURATION_FILE}" ]; then
+        echo "Configuration file not found or readable: $CONFIGURATION_FILE"
+        return 1
+    fi
+
+    while IFS='=' read -r key value; do
+        # Skip empty lines and comments.
+        [[ -z "$key" || "$key" =~ ^[[:space:]]*# ]] && continue
+
+        # Remove leading and trailing spaces.
+        key="${key#"${key%%[![:space:]]*}"}"
+        value="${value#"${value%%[![:space:]]*}"}"
+        key="${key%"${key##*[![:space:]]}"}"
+        value="${value%"${value##*[![:space:]]}"}"
+
+        # Remove surrounding double and single quotes on values, if fully quoted.
+        if [[ "$value" =~ ^\".*\"$ ]]; then
+            value="${value:1:-1}"
+        elif [[ "$value" =~ ^\'.*\'$ ]]; then
+            value="${value:1:-1}"
+        fi
+
+        # Ensure key is valid.
+        case "$key" in
+            AZURE_EPHEMERAL_DISK_SETUP_AGGREGATION|\
+            AZURE_EPHEMERAL_DISK_SETUP_FS_TYPE|\
+            AZURE_EPHEMERAL_DISK_SETUP_MDADM_CHUNK|\
+            AZURE_EPHEMERAL_DISK_SETUP_MDADM_NAME|\
+            AZURE_EPHEMERAL_DISK_SETUP_MOUNT_POINT|\
+            AZURE_EPHEMERAL_DISK_SETUP_SCSI_RESOURCE|\
+            AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS)
+                ;;
+            *)
+                echo "WARNING: ignoring invalid config key=$key" >&2
+                continue
+                ;;
+        esac
+
+        # Value will be checked later, just assume it is correct and export it.
+        echo "Configuration file set key=$key with value=$value" >&2
+        export "$key=$value"
+    done < "$CONFIGURATION_FILE"
+}
+
+# Wait for a systemd unit to become active for up to 60 (AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS) seconds.
+start_and_wait_for_systemd_unit() {
+    local unit="$1"
+    local deadline=$((SECONDS + AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS))
+
+    if ! systemctl is-enabled "$unit" >&2; then
+        log "systemd unit $unit is not enabled, skipping wait"
+        return 1
+    fi
+
+    # systemctl --wait may unexpectedly hang in some scenarios, so wait for up to 1s and loop as needed.
+    while [[ $SECONDS -lt $deadline ]]; do
+        log "Waiting for systemd unit $unit to become active... (remaining: $((deadline - SECONDS)) seconds)"
+        timeout 1s systemctl start "$unit" --wait >&2 || log "Failed to start systemd unit $unit: $?"
+
+        if systemctl is-active "$unit" >&2; then
+            log "Started systemd unit $unit successfully"
+            return 0
+        fi
+    done
+
+    log "Timed out waiting for systemd unit $unit to become active: $(systemctl status "$unit" 2>&1 || true)"
+    return 1
+}
+
+wait_for_udev_to_settle() {
+    log "Waiting for udev to settle..."
+    udevadm settle --timeout 30 || true
```

Comment:
Is 30s enough for a very big sku? I feel like we might need some retry here and allow for some control of this value or how many retries. This is probably not a problem with distros using cloud-init because by now cloud-init has already done the ephemeral networking and it has already done the udevadm settle earlier. However, it might be a problem with distros that don't use cloud-init

Reply:
added undocumented AZURE_EPHEMERAL_DISK_SETUP_UDEVADM_SETTLE_TIMEOUT_SECS

Reply:
and upped the default to 60s